### PR TITLE
Cv 204 Web UI changes add another apprentice to a cohort

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System.ComponentModel.DataAnnotations;
+using SFA.DAS.Authorization;
+
+namespace SFA.DAS.ProviderCommitments.Web.Models
+{
+    public class Cohort : IAuthorizationContextModel
+    {
+        [Required]
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs]: It looks like fluent validation has been included in the Mvc pipeline so the validation attributes shouldn't be needed here.

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Requests/AddDraftApprenticeshipRequest.cs [hunk 2]
```diff
@@ -1,11 +1,13 @@
 ﻿using System;
+using System.ComponentModel.DataAnnotations;
 using SFA.DAS.Authorization;
 
 namespace SFA.DAS.ProviderCommitments.Web.Requests
 {
     public class AddDraftApprenticeshipRequest : IAuthorizationContextModel
     {
-        public Guid ReservationId { get; set; }
+        [Required]
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Requests/AddDraftApprenticeshipRequest.cs]: Same as above.

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+﻿using System.ComponentModel.DataAnnotations;
+using SFA.DAS.Authorization;
+
+namespace SFA.DAS.ProviderCommitments.Web.Models
+{
+    public class Cohort : IAuthorizationContextModel
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs]: This should probably be suffixed with `Request` and moved to the `SFA.DAS.ProviderCommitments.Web.Requests` namespace to follow the project conventions. It also looks like we currently have three types that are inputs to the various `AddDraftApprenticeship` action methods:

1. `Cohort`
2. `AddDraftApprenticeship`
3. `ReservationsAddDraftApprenticeship`

What do you think to renaming these to something along the lines of:

1. `NonReservationsAddDraftApprenticeship`
2. `CreateCohortReuqest`
3. `ReservationsAddDraftApprenticeship`

(1 & 3 may also be able to be consolidated, see comment above).

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            _providerCommitmentsService = providerCommitmentsService;
+            _addDraftApprenticeshipToCohortRequestMapper = addDraftApprenticeshipToCohortRequestMapper;
+            _urlHelper = urlHelper;
+        }
+
+        [HttpGet]
+        [Route("add-apprentice")]
+        public async Task<IActionResult> AddDraftApprenticeship(Cohort cohort)
+        {
+            if (!ModelState.IsValid)
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs]: I don't think would work as we need to deal with ViewModels having a bad ModelState (on PostBack) and send the user to a particular page

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            _mediator = mediator;
+            _providerCommitmentsService = providerCommitmentsService;
+            _addDraftApprenticeshipToCohortRequestMapper = addDraftApprenticeshipToCohortRequestMapper;
+            _urlHelper = urlHelper;
+        }
+
+        [HttpGet]
+        [Route("add-apprentice")]
+        public async Task<IActionResult> AddDraftApprenticeship(Cohort cohort)
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs]: Yes they could, but this makes the controller action quite complicated (having to deal with both situations). Also the endpoint without the reservation information will be removed from the system in the near future (once we turn reservations on fully)

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                CohortPublicHashedId = request.CohortPublicHashedId,
+                CohortId = request.CohortId,
+                StartDate = new MonthYearModel(request.StartMonthYear),
+                ReservationId = request.ReservationId,
+                CourseCode = request.CourseCode
+            };
+
+            await AddLegalEntityAndCoursesToModel(model);
+
+            return View("AddDraftApprenticeship", model);
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs]: removed

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs (hunk 3 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs]: I agree with the naming. But I think 2 should be renamed to 'CreateCohortWithDraftApprenticeshipRequest' as there

Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs]: Also I've renamed the controller to a matching name 'CreateCohortWithDraftApprenticeshipController'

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs (hunk 1 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Models/Cohort.cs]: Replaced with Fluent validation

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Requests/AddDraftApprenticeshipRequest.cs (hunk 2 above)
Comment (Pauljgraham) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Requests/AddDraftApprenticeshipRequest.cs]: Done

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs]: Yeah, we could update the `SFA.DAS.Validation.Mvc` package to support .NET Core to allow this. Let's leave it as is for now though.

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/DraftApprenticeshipController.cs]: :+1:

**Pull request closed without merging**